File: 2023/day05_1.py
# pylint: disable=missing-module-docstring, missing-function-docstring

from typing import Any

type Table = tuple[int, int, int]


def solution(content: list[str]) -> Any:
    maps = parse(content[2:])
    seeds = parse_seeds(content[0])
    return min(convert(seed, maps) for seed in seeds)


def parse_seeds(line: str):
    return [int(x) for x in line[7:].split(" ")]


def parse(lines: list[str]) -> list[list[Table]]:
    output: list[list[Table]] = []
    for line in lines:
        if line.strip() == "":
            continue
        elif "-" in line:
            output.append([])
        else:
            output = output[:-1] + [output[-1] + [parse_one(line)]]
    return output


def parse_one(line: str) -> Table:
    [dst, src, size] = [int(x) for x in line.split(" ")]
    return (dst, src, size)


def convert(seed: int, maps: list[list[Table]]) -> int:
    if not maps:
        return seed
    [fst, *rest] = maps
    for dst, src, size in fst:
        if src <= seed < src + size:
            return convert(dst + seed - src, rest)
    return convert(seed, rest)


def main() -> None:
    with open("day05.dat", "r", encoding="utf8") as file:
        print(solution(file.readlines()))


if __name__ == "__main__":
    main()

File: 2023/day05-1.ts
type Table = {
  dst: number;
  src: number;
  size: number;
};

function solution(content: string) {
  const lines = content.split("\n");
  const maps = parse(lines.slice(2));
  const seeds = parseSeeds(lines[0]);

  return Math.min(...seeds.map((seed) => convert(seed, maps)));
}

function parseSeeds(line: string) {
  return line.slice(7).split(" ").map(Number);
}

function parse(lines: string[]): Table[][] {
  const output: Table[][] = [];
  lines.forEach((line) => {
    if (line.trim() === "") return;
    if (line.includes("-")) output.push([]);
    else output.at(-1)?.push(parseOne(line));
  });
  return output;
}

function parseOne(line: string): Table {
  const values = line.split(" ").map(Number);
  return {
    dst: values[0],
    src: values[1],
    size: values[2],
  };
}

function convert(seed: number, maps: Table[][]): number {
  if (maps.length === 0) return seed;
  const [fst, ...rest] = maps;
  const map = fst.find(({ src, size }) => src <= seed && seed < src + size);
  if (map) return convert(map.dst + seed - map.src, rest);
  return convert(seed, rest);
}

Deno.readTextFile("day05.dat")
  .then((content) => content.trim())
  .then(solution)
  .then(console.log);
